File: frontend/src/rules/effects.test.ts
import { describe, expect, it } from 'vitest';
import { createLevel3Character } from './characters';
import { applyDamage, applyDmEffects, resolveShortRest, resolveSpellEffect } from './effects';

describe('automatic effect settlement', () => {
  it('uses temporary hp before regular hp', () => {
    const result = applyDamage({ id: 'x', name: 'x', side: 'enemy', initiativeBonus: 0, initiative: 0, ac: 10, hp: 10, maxHp: 10, temporaryHp: 5, attackBonus: 0, damage: '1d4', damageType: '鈍擊' }, 8);
    expect(result).toMatchObject({ temporaryHp: 0, hp: 7 });
  });

  it('automatically spends hit dice on a short rest until full', () => {
    const player = { ...createLevel3Character('player1', '甲', '戰士'), hp: 1 };
    const result = resolveShortRest(player, () => 0.99);
    expect(result.character.hp).toBe(player.maxHp);
    expect(result.diceSpent).toBeGreaterThan(0);
  });

  it('settles healing spells and bounds DM damage', () => {
    const caster = createLevel3Character('player1', '牧者', '牧師');
    const wounded = { ...createLevel3Character('player2', '傷者', '戰士'), hp: 1 };
    const healed = resolveSpellEffect([caster, wounded], undefined, caster.id, wounded.id, { kind: 'healing', target: 'ally', dice: '2d4', flat: 2 }, () => 0);
    expect(healed.players[1].hp).toBe(5);
    const dm = applyDmEffects(healed.players, [{ targetId: 'player2', kind: 'damage', amount: 9999, reason: '落石' }]);
    expect(dm.players[1].hp).toBe(0);
    expect(dm.logs[0]).toMatch(/落石/);
  });

  it('keeps combat hp synchronized after healing and consumes temporary hp for DM damage', () => {
    const caster = createLevel3Character('player1', '牧者', '牧師');
    const wounded = { ...createLevel3Character('player2', '傷者', '戰士'), hp: 1, temporaryHp: 3 };
    const combat = {
      active: true, round: 1, turnIndex: 0,
      combatants: [
        { id: caster.id, playerId: caster.id, name: caster.name, side: 'party' as const, initiativeBonus: 0, initiative: 20, ac: caster.ac, hp: caster.hp, maxHp: caster.maxHp, attackBonus: 0, damage: '1d4', damageType: '鈍擊' },
        { id: wounded.id, playerId: wounded.id, name: wounded.name, side: 'party' as const, initiativeBonus: 0, initiative: 10, ac: wounded.ac, hp: wounded.hp, maxHp: wounded.maxHp, temporaryHp: 3, attackBonus: 0, damage: '1d4', damageType: '鈍擊' },
      ],
    };
    const healed = resolveSpellEffect([caster, wounded], combat, caster.id, wounded.id, { kind: 'healing', target: 'ally', flat: 4 }, () => 0);
    expect(healed.combat?.combatants[1].hp).toBe(5);
    const dm = applyDmEffects(healed.players, [{ targetId: 'player2', kind: 'damage', amount: 5, reason: '落石' }]);
    expect(dm.players[1]).toMatchObject({ temporaryHp: 0, hp: 3 });
  });

  it('uses the visible spell attack total instead of rolling a hidden d20', () => {
    const caster = createLevel3Character('player1', '術者', '術士');
    const enemy = { id: 'enemy', name: '敵人', side: 'enemy' as const, initiativeBonus: 0, initiative: 10, ac: 15, hp: 10, maxHp: 10, attackBonus: 0, damage: '1d4', damageType: '鈍擊' };
    const combat = { active: true, round: 1, turnIndex: 0, combatants: [{ ...enemy }] };
    const missed = resolveSpellEffect([caster], combat, caster.id, enemy.id, { kind: 'damage', target: 'creature', flat: 4, attackRoll: true }, () => .99, { attackTotal: 14 });
    expect(missed.amount).toBe(0);
    expect(missed.combat?.combatants[0].hp).toBe(10);
    const hit = resolveSpellEffect([caster], combat, caster.id, enemy.id, { kind: 'damage', target: 'creature', flat: 4, attackRoll: true }, () => 0, { attackTotal: 15 });
    expect(hit.combat?.combatants[0].hp).toBe(6);
  });
});

File: backend/internal/rules/effects_test.go
package rules

// Ported 1:1 from frontend/src/rules/effects.test.ts ("automatic effect
// settlement"), preserving every vitest assertion.

import (
	"strings"
	"testing"
)

// mustResolveSpellEffect mirrors the tests calling resolveSpellEffect
// directly (any throw fails the vitest test).
func mustResolveSpellEffect(t *testing.T, players []Character, combat *CombatState, casterID, targetID string, effect SpellEffect, random RandomSource, forced *ForcedRolls) SpellEffectResult {
	t.Helper()
	result, err := ResolveSpellEffect(players, combat, casterID, targetID, effect, random, forced)
	if err != nil {
		t.Fatalf("resolveSpellEffect threw: %v", err)
	}
	return result
}

func TestUsesTemporaryHpBeforeRegularHp(t *testing.T) {
	result := ApplyDamage(Combatant{ID: "x", Name: "x", Side: "enemy", InitiativeBonus: 0, Initiative: 0, AC: 10, HP: 10, MaxHP: 10, TemporaryHP: 5, AttackBonus: 0, Damage: "1d4", DamageType: "鈍擊"}, 8)
	if result.TemporaryHP != 0 {
		t.Fatalf("expected temporaryHp 0, got %d", result.TemporaryHP)
	}
	if result.HP != 7 {
		t.Fatalf("expected hp 7, got %d", result.HP)
	}
}

func TestAutomaticallySpendsHitDiceOnShortRestUntilFull(t *testing.T) {
	player := CreateLevel3Character("player1", "甲", "戰士")
	player.HP = 1
	result := ResolveShortRest(player, func() float64 { return 0.99 })
	if result.Character.HP != player.MaxHP {
		t.Fatalf("expected hp %d, got %d", player.MaxHP, result.Character.HP)
	}
	if !(result.DiceSpent > 0) {
		t.Fatalf("expected diceSpent > 0, got %d", result.DiceSpent)
	}
}

func TestSettlesHealingSpellsAndBoundsDmDamage(t *testing.T) {
	caster := CreateLevel3Character("player1", "牧者", "牧師")
	wounded := CreateLevel3Character("player2", "傷者", "戰士")
	wounded.HP = 1
	healed := mustResolveSpellEffect(t, []Character{caster, wounded}, nil, caster.ID, wounded.ID, SpellEffect{Kind: "healing", Target: "ally", Dice: "2d4", Flat: 2}, func() float64 { return 0 }, nil)
	if healed.Players[1].HP != 5 {
		t.Fatalf("expected healed hp 5, got %d", healed.Players[1].HP)
	}
	dmPlayers, dmLogs := ApplyDmEffects(healed.Players, []DMEffect{{TargetID: "player2", Kind: "damage", Amount: 9999, Reason: "落石"}})
	if dmPlayers[1].HP != 0 {
		t.Fatalf("expected dm-damaged hp 0, got %d", dmPlayers[1].HP)
	}
	if len(dmLogs) == 0 || !strings.Contains(dmLogs[0], "落石") {
		t.Fatalf("expected first log to match 落石, got %v", dmLogs)
	}
}

func TestKeepsCombatHpSynchronizedAfterHealingAndConsumesTemporaryHpForDmDamage(t *testing.T) {
	caster := CreateLevel3Character("player1", "牧者", "牧師")
	wounded := CreateLevel3Character("player2", "傷者", "戰士")
	wounded.HP = 1
	wounded.TemporaryHP = 3
	combat := &CombatState{
		Active: true, Round: 1, TurnIndex: 0,
		Combatants: []Combatant{
			{ID: caster.ID, PlayerID: caster.ID, Name: caster.Name, Side: "party", InitiativeBonus: 0, Initiative: 20, AC: caster.AC, HP: caster.HP, MaxHP: caster.MaxHP, AttackBonus: 0, Damage: "1d4", DamageType: "鈍擊"},
			{ID: wounded.ID, PlayerID: wounded.ID, Name: wounded.Name, Side: "party", InitiativeBonus: 0, Initiative: 10, AC: wounded.AC, HP: wounded.HP, MaxHP: wounded.MaxHP, TemporaryHP: 3, AttackBonus: 0, Damage: "1d4", DamageType: "鈍擊"},
		},
	}
	healed := mustResolveSpellEffect(t, []Character{caster, wounded}, combat, caster.ID, wounded.ID, SpellEffect{Kind: "healing", Target: "ally", Flat: 4}, func() float64 { return 0 }, nil)
	if healed.Combat == nil || healed.Combat.Combatants[1].HP != 5 {
		t.Fatalf("expected combat combatant hp 5, got %+v", healed.Combat)
	}
	dmPlayers, _ := ApplyDmEffects(healed.Players, []DMEffect{{TargetID: "player2", Kind: "damage", Amount: 5, Reason: "落石"}})
	if dmPlayers[1].TemporaryHP != 0 {
		t.Fatalf("expected temporaryHp 0, got %d", dmPlayers[1].TemporaryHP)
	}
	if dmPlayers[1].HP != 3 {
		t.Fatalf("expected hp 3, got %d", dmPlayers[1].HP)
	}
}

func TestUsesTheVisibleSpellAttackTotalInsteadOfRollingAHiddenD20(t *testing.T) {
	caster := CreateLevel3Character("player1", "術者", "術士")
	enemy := Combatant{ID: "enemy", Name: "敵人", Side: "enemy", InitiativeBonus: 0, Initiative: 10, AC: 15, HP: 10, MaxHP: 10, AttackBonus: 0, Damage: "1d4", DamageType: "鈍擊"}
	combat := &CombatState{Active: true, Round: 1, TurnIndex: 0, Combatants: []Combatant{enemy}}
	attackTotal14 := 14
	missed := mustResolveSpellEffect(t, []Character{caster}, combat, caster.ID, enemy.ID, SpellEffect{Kind: "damage", Target: "creature", Flat: 4, AttackRoll: true}, func() float64 { return 0.99 }, &ForcedRolls{AttackTotal: &attackTotal14})
	if missed.Amount != 0 {
		t.Fatalf("expected missed amount 0, got %d", missed.Amount)
	}
	if missed.Combat == nil || missed.Combat.Combatants[0].HP != 10 {
		t.Fatalf("expected missed combatant hp 10, got %+v", missed.Combat)
	}
	attackTotal15 := 15
	hit := mustResolveSpellEffect(t, []Character{caster}, combat, caster.ID, enemy.ID, SpellEffect{Kind: "damage", Target: "creature", Flat: 4, AttackRoll: true}, func() float64 { return 0 }, &ForcedRolls{AttackTotal: &attackTotal15})
	if hit.Combat == nil || hit.Combat.Combatants[0].HP != 6 {
		t.Fatalf("expected hit combatant hp 6, got %+v", hit.Combat)
	}
}
